patches/server/0925-Paper-Plugins.patch

```diff
@@ -0,0 +1,1274 @@
++                case INITIALIZED -> NamedTextColor.GREEN;
++                case ERRORED -> NamedTextColor.RED;
++            };
++        } else if (provider instanceof PaperPluginProvider.PaperServerPluginProvider serverPluginProvider && serverPluginProvider.shouldSkipCreation()) {
++            // Paper plugins will be skipped if their provider is skipped due to their initializer failing.
++            // Show them as red
++            return NamedTextColor.RED;
++        } else {
++            return NamedTextColor.GRAY;
++        }
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/provider/PluginProviders.java b/src/main/java/io/papermc/paper/plugin/provider/PluginProviders.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..277e8b3802d5cb07893be91a27f31b8a349fc691
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/PluginProviders.java
+@@ -0,0 +1,67 @@
++package io.papermc.paper.plugin.provider;
++
++import io.papermc.paper.plugin.provider.type.PluginFileType;
++import org.slf4j.Logger;
++import org.slf4j.LoggerFactory;
++
++import java.io.IOException;
++import java.nio.file.Files;
++import java.nio.file.Path;
++import java.util.jar.JarFile;
++
++/*
++Responsible for storing all the plugin files.
++ */
++public class PluginProviders {
++
++    public static final PluginProviders INSTANCE = new PluginProviders();
++    private static final Logger LOGGER = LoggerFactory.getLogger(PluginProviders.class);
++
++    public void handlePluginDirectory(Path toPath) throws IOException {
++        if (!Files.isDirectory(toPath)) { // Create if missing
++            Files.createDirectories(toPath);
++        }
++
++        this.addDirectory(toPath);
++    }
++
++    public void addDirectory(Path directory) {
++        try {
++            Files.walk(directory, 1).filter(Files::isRegularFile).forEach(this::addFile);
++        } catch (Exception e) {
++            throw new RuntimeException(e);
++        }
++    }
++
++    // TODO: This will log invalid plugins in the directory, i'm guessing this wasn't the behavior before.
++    // TODO: Update this to properly do better warnings, whatever
++    public void addFile(Path source) {
++        if (Files.notExists(source)) {
++            LOGGER.warn("File '{}' specified through 'add-plugin' argument does not exist, cannot load a plugin from it!", source);
++            return;
++        }
++
++        if (!Files.isRegularFile(source)) {
++            LOGGER.warn("File '{}' specified through 'add-plugin' argument is not a file, cannot load a plugin from it!", source);
++            return;
++        }
++
++        if (!source.getFileName().toString().endsWith(".jar")) {
++            LOGGER.warn("File '{}' specified through 'add-plugin' argument is not a jar file, cannot load a plugin from it!", source);
++            return;
++        }
++
++        try {
++            JarFile file = new JarFile(source.toFile());
```

Comment:
Is this zip file is correctly closed ?

Reply:
There isn't proper logic for closing these, but because of how plugin providers work they are meant to pretty much always be sitting around.

Reply:
I mean when the plugin is unloaded for example ?

Reply:
Because plugins are separate from their loaders, it can't be certain that we want to actually throw away this jar anymore.